Prepare for upgrade guide for CuPy v14 (#9485)
cupy/cupy: docs/source/upgrade.rst

> Support for NVIDIA CUDA component wheels

Reviewer: Just to not have it on the main thread, maybe something like:
```
Minimal support for structured dtypes
-------------------------------------
CuPy now accepts most structured dtypes when converting from NumPy or
creating a new array with ``empty`` and ``zeros``.
However, the only supported operation on these is accessing a single field::

    arr[field_name]
    arr[field_name] = value

As of v14.0 any kernel launch with a structured dtype will fail
(including ``arr.copy()``).

Note that ``arr[field_name]`` must be sufficiently aligned for the GPU.
This is not guaranteed by NumPy even with ``align=True`` as the GPU has
larger alignment requirements.
```

----------------------------------------

CuPy v14 can be installed together with a minimal CUDA installation all from PyPI (ex: ``pip install "cupy-cuda13x[ctk]"``),
allowing quickly spinning up a fresh virtual environment without any system-wide CUDA Toolkit pre-installed (only the CUDA driver is needed),

> thereby enabling smaller installation footprint and better interoperability with other Python GPU libraries. See :doc:`install` for details.

Reviewer: Just suggestions, take or leave
```suggestion
CuPy v14 can be installed together with a minimal CUDA installation from PyPI (ex: ``pip install "cupy-cuda13x[ctk]"``),
allowing to quickly spin up a fresh virtual environment without a pre-installed CUDA Toolkit (only the CUDA driver is needed).
This enables a smaller installation footprint and better interoperability with other Python GPU libraries. See :doc:`install` for details.
```

* ``cupy.cusparse`` -> :mod:`cupyx.scipy.sparse.cusparse` (undocumented API, deprecated in CuPy v12)
* ``cupy.cutensor`` -> :mod:`cupyx.scipy.linalg.cutensor` (undocumented API, deprecated in CuPy v12)

Requirement Changes
-------------------

The following versions are no longer supported in CuPy v14.

* CUDA 11.x or earlier (CUDA 12.0 or later is now required)
* Python 3.9 or earlier (Python 3.10 or later is now required)
* NumPy 1.x (NumPy 2.0 or later is now required)

> * ROCm 6.x or earlier (ROCm 7.0 or later is now required)

Reviewer 2: ROCm 6 will still build and test around the same as ROCm 7. Is this the same for the other deps listed here (CUDA, Python, NumPy)? If so, maybe it would be better to say something like "test support has dropped for the following versions"

Reviewer 3: In CuPy, currently there are only two possible library support statuses: either "supported" (meaning the configuration is tested in CI) or "not supported" (not tested in CI, with potential of the code paths for the version of that dep library being removed).

Considering the current situation and "experimental" status of the ROCm support in CuPy, it would be preferable not to mention ROCm/RCCL/... in the upgrade guide.

> * NCCL 2.17 or earlier (NCCL 2.18 or later is now required)

Reviewer 2: Can we mention RCCL 😃

     - Docs
   * - v14
     - 3.5~
     - 12.0~
     - 7.0~

>      - 2.0~

Author: ```suggestion
     - 2.3~
```

     - 2.18~
     - n/a
     - 3.10~
     - 2.0~
     - 1.7~

>      - NumPy 1.26 & SciPy 1.11

Author: @kmaehashi is this correct?
```suggestion
     - NumPy 2.0 & SciPy 1.11
```

Reviewer 3: We use NumPy v2.3 and SciPy v1.16 as baseline.

https://docs.cupy.dev/en/latest/install.html#python-dependencies

```suggestion
     - NumPy 2.3 & SciPy 1.16
```

     - Docs
   * - v14

>      - 3.5~

Reviewer 3: Minimum CC supported is now 5.0 as we dropped CUDA 11.

```suggestion
     - 5.0~
```
 https://github.com/cupy/cupy/blob/d99df5b322d948b85bc1da486da893ad56fe6c77/install/cupy_builder/_compiler.py#L102

     - 12.0~
     - 7.0~
     - 2.3~
     - 2.18~
     - n/a
     - 3.10~
     - 2.0~

>      - 1.7~

Reviewer 3: ```suggestion
     - 1.14~
```
